pkg/util/taskgroup/taskgroup.go

```diff
@@ -0,0 +1,109 @@
+// Originally taken from https://github.com/ptxmac/multierrgroup
+
+// Package taskgroup provides a mix of multierr and errgroup
+// See documentation for https://pkg.go.dev/go.uber.org/multierr and https://pkg.go.dev/golang.org/x/sync/errgroup
+package taskgroup
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+)
+
+// Group manages goroutines and collect all the errors.
+// See https://pkg.go.dev/golang.org/x/sync/errgroup#group for more information
+type Group interface {
+	WithContext(ctx context.Context) context.Context
+	WithPanicHandler(f func(any))
+	Wait() error
+	Go(name string, f func(logger *zap.Logger) error)
+}
+
+type group struct {
+	cancel       context.CancelFunc
+	logger       *zap.Logger
+	panicHandler func(any)
+
+	wg sync.WaitGroup
+
+	errMutex sync.Mutex
+	err      error
+}
+
+// NewGroup returns a new Group.
+func NewGroup(logger *zap.Logger) Group {
+	return &group{
+		cancel:       nil, // Set separately by WithContext
+		panicHandler: nil, // Set separately by WithPanicHandler
+		logger:       logger,
+		wg:           sync.WaitGroup{},
+
+		errMutex: sync.Mutex{},
+		err:      nil,
+	}
+}
+
+// WithContext updates the current Group with a associated Context.
+// The context will be canceled if any goroutine returns an error.
+// See https://pkg.go.dev/golang.org/x/sync/errgroup#WithContext
+func (g *group) WithContext(ctx context.Context) context.Context {
+	ctx, g.cancel = context.WithCancel(ctx)
+	return ctx
+}
+
+// WithPanicHandler sets a panic handler for the group.
+func (g *group) WithPanicHandler(f func(any)) {
```

Comment:
Is `WithPanicHandler` currently used anywhere? (couldn't find anything, but maybe I missed it)

Reply:
Ah I see. Do you plan to port that to taskgroup soon after? (generally trying to avoid adding dead code -- alternatively, how much effort would it be to add this back when needed?)

My 2 cents - porting the autoscaler-agent's usage is probably a bit more complicated, and this work is not super high priority, so it's _probably_ a safer bet to keep it minimal for now.

WDYT?

Reply:
> Do you plan to port that to taskgroup soon after? 

I didn't have speific plans.

> generally trying to avoid adding dead code

Generally, yes, but in this case it is a "general case" abstraction, so it's design doesn't have to strictly follow whether it is used in the codebase. Also, the change is tiny, shouldn't be in the way.

Reply:
Fair enough, makes sense